include/aws/common/file.h

```diff
@@ -4,18 +4,113 @@
  * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
  * SPDX-License-Identifier: Apache-2.0.
  */
+#include <aws/common/byte_buf.h>
 #include <aws/common/common.h>
 
+#ifdef _WIN32
+#    define AWS_PATH_DELIM '\\'
+#    define AWS_PATH_DELIM_STR "\\"
+#else
+#    define AWS_PATH_DELIM '/'
+#    define AWS_PATH_DELIM_STR "/"
+#endif
+
+struct aws_string;
+
+enum aws_file_type {
+    AWS_FILE_TYPE_FILE = 1,
+    AWS_FILE_TYPE_SYM_LINK = 2,
+    AWS_FILE_TYPE_DIRECTORY = 4,
+};
+
+struct aws_directory_entry {
+    /**
+     * Absolute path to the entry from the current process root.
+     */
+    struct aws_byte_cursor path;
+    /**
+     * Path to the entry relative to the current working directory.
+     */
+    struct aws_byte_cursor relative_path;
+    /**
+     * Bit-field of enum aws_file_type
+     */
+    int file_type;
+    /**
+     * Size of the file on disk.
+     */
+    int64_t file_size;
+};
+
+/**
+ * Invoked during calls to aws_directory_traverse() as an entry is encountered. entry will contain
+ * the parsed directory entry info.
+ *
+ * Return true to continue the traversal, or alternatively, if you have a reason to abort the traversal, return false.
+ */
+typedef bool(aws_on_directory_entry)(const struct aws_directory_entry *entry, void *user_data);
+
 AWS_EXTERN_C_BEGIN
 
 /**
- * To support non-ascii file path across platform.
- * For windows, _wfopen will be invoked under the hood. For other platforms, same as calling fopen
- * Functionality is the same as fopen.
- * On error, errno will be set, and NULL will be returned. Same as fopen.
+ * Don't use this. It never should have been added in the first place. It's now deprecated.
+ */
+AWS_COMMON_API FILE *aws_fopen(const char *file_path, const char *mode);
```

Comment:
Why don't just completely remove it?

Reply:
That would be an API break. Instead, we need to deprecate it and remove it later.